### DISCUSSION TITLE
field name/value in error message

### DISCUSSION THREAD

**Discussion by teathinker** [simdjson/simdjson#1699] [Ideas]
Hi, sometimes I receive field-related exception such as INCORRECT_TYPE, NO_SUCH_FIELD. I am wondering if we can add field name into error message otherwise it would take peopple more time to debug code and find out. Thanks.

**Comment by lemire** [comment]
The library should return NO_SUCH_FIELD following a query such as obj["myfied"] which should not be hard to debug.

If you have some specific issue, please elaborate. Consider providing code samples.

The error codes in simdjson are part of an enums. They are fixed size and allocation free They do not and cannot capture the context of use.

So it is important for you to really explained the issue with precision so we can see how we can help.

**Reply by teathinker** [reply]
Sometimes I write parsing code next to each other, for example:

```
auto height = doc["data"]["height"].get_double().value();
auto weight = doc["data"]["weight"].get_double().value();
...
auto first_name = doc["data"]["first_name"].get_string().value();
auto last_name = doc["data"]["last_name"].get_string().value();
```

There could be 10+ fieilds to read. However if I get a INCORRECT_TYPE exception on production, I couldn't figure out which field it's talking about so I have to either log the json or setup debugging. Do you have better way? 

I definitely can try to solve the issue during dev testing, but as you know, client sides can wire whatever json on production so it's still better to have a message indicating which field/value is wrong. 

I read the code and see the error messages are pretty static, but there could be a way to put field/value infomation into the exception so what() method can show it. Exception is already expensive so building a string isn't expected to add much.

Thanks.

**Reply by lemire** [reply]
I would discourage you from writting the code in this manner. It is probably quite inefficient (computationally). You are constantly looking up the field "data". Don't do that!

You should do...
```
ondemand::object dataobject;
error = doc["data"].get(dataobject);
if(error) {...}
```

Exceptions make for clean code, but they also make it harder to have precise error control. Please consider using the approach without exceptions... it is more verbose, but it will give you tight control...

```
error = dataobject["height"].get(val);
if(error) {...}
error = val.get_double().get(height);
if(error) {...}
```

and so forth. At each step you can include the precise error handling you need, instead of relying on a crude try/catch.

**Reply by teathinker** [reply]
Thanks for your advice on performance.
I do see the way to handle error after each read without expecting an expcetion. It's just a little trivial to me given there are 10+ feilds there to handle, and all the code can do is still throwing an exception with field/value information. Maybe I can use an inline function to save some effort.

**Reply by lemire** [reply]
You definitely can use your own functions to reduce the programming burden and still have the exact error handling you seek.

**Reply by lemire** [reply]
Note that I take very seriously feedback such as yours.

**Comment by ghost** [comment]
May be an optional helper class to collect errors is a solution. I just created an interface to `simdjson`  with an prototype approach (see [simdjson_peval](https://github.com/kla-sch/simdjson_peval#readme)). I have created an error collector, which is able to trace paths to error locations:

```c++
simdjson_peval::error error_collection(true /* switch path collection on */);
{
    simdjson_peval::error::path_scope hight_scope(&error_collection, "height");
    auto error = dataobject["height"].get(val);
    if (error) {
        error_collection.add(error);
    }
}
...

if (error_collection) {
    std::err << error_collection.to_string();
    // output: at <root>.height: <the error message>
}
```

See also: [UserGuide/Save Errors](https://github.com/kla-sch/simdjson_peval/blob/main/doc/UserGuide.md#save-errors)

May such a class could also be integrated somehow:
```c++
dataobject["height"].get_collect_error(val, error_container, "height");
// or
arrayElement.get_collect_error(val, error_container, arrayIndex);
```

**Reply by lemire** [reply]
Wow.

Your project (simdjson_peval) looks very interesting.

cc @jkeiser
